### ISSUE TITLE
Plotting Issue (Online Branch)

### ISSUE THREAD

**Issue by samuel-marsh** [welch-lab/liger#166]
Hi,

Wanted to post issue I'm having with online branch so hopefully can be addressed before release.  I'll detail the object setup and issue below but in short many of the plotting functions from the online branch are not working on my liger object.  However, if I take code of functions as they are currently written in master branch and set them up as new functions those functions work on new object with no issues (as well as the modified code from new PR I submitted the other day).  So something appears to be happening as result of changes in code for some of the online branch plot functions.

Let me know if you have any other questions about issue/errors I'm having.

Thanks!
Sam

**Object setup and analysis**
I'm doing an analysis of liger object that was created by first importing and QC filtering data in Seurat V3, then converting from Seurat to liger object with `combined = TRUE`.

**Then I'm running online_iNMF with the following commands:**
```
object_name <-  normalize(object_name)
object_name <-  selectGenes(object_name, do.plot = T, num.genes = 900)
object_name <-  scaleNotCenter(object_name)
object_name <-  online_iNMF(object_name, k = 40, lambda = 10, miniBatch_size = 2000)
object_name <-  quantile_norm(object_name)
object_name <-  clusterLouvainJaccard(object_name,resolution = 0.5)
object_name <-  runUMAP(object_name)
```
*I switch to Seurat 2.3.4 during this process as clusterlouvainjaccard is setup for Seurat V2 object structure.

If I run `plotByDatasetAndCluster` that works with no issues.

**However, when I try and use plotGene:**
```
> p1 <- plotGene(object = object_name, gene = "FOS", return.plots = T)
[1] X1 X2
<0 rows> (or 0-length row.names)
Error in `$<-.data.frame`(`*tmp*`, "plotby", value = c(3L, 3L, 3L, 3L,  : 
  replacement has 105395 rows, data has 0
In addition: Warning messages:
1: In min(dr_df$dr1) : no non-missing arguments to min; returning Inf
2: In max(dr_df$dr1) : no non-missing arguments to max; returning -Inf
3: In min(dr_df$dr2) : no non-missing arguments to min; returning Inf
4: In max(dr_df$dr2) : no non-missing arguments to max; returning -Inf
```

**plotGeneViolin also has issue:**
```
> p2 <- plotGeneViolin(object = leng_combined, gene = "FOS", return.plots = T)
Error in object@h5file.info[[1]] : subscript out of bounds
```

However, if I go to the current liger master branch and copy to the plotGene and plotGeneViolin function code and set them up as new functions (with new names) and then use them on the object the plots work with no issues.  So something appears to be happening as result of changes in code for online branch plotting.

**Package loading and version info:**
```
library(tidyverse)
library(Seurat)
library(liger)
```
Here is sessionInfo() output:
```
> sessionInfo()
R version 3.5.3 (2019-03-11)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 9 (stretch)

Matrix products: default
BLAS: /usr/lib/libblas/libblas.so.3.7.0
LAPACK: /usr/lib/lapack/liblapack.so.3.7.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8    LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] liger_0.5.0.9000  data.table_1.12.8 hdf5r_1.3.2.9000  patchwork_1.0.0   Seurat_2.3.4      Matrix_1.2-18    
 [7] cowplot_1.0.0     forcats_0.5.0     stringr_1.4.0     dplyr_0.8.5       purrr_0.3.4       readr_1.3.1      
[13] tidyr_1.1.0       tibble_3.0.1      ggplot2_3.3.0     tidyverse_1.3.0  

loaded via a namespace (and not attached):
  [1] readxl_1.3.1        snow_0.4-3          backports_1.1.7     Hmisc_4.4-0         igraph_1.2.5       
  [6] sn_1.6-1            plyr_1.8.6          splines_3.5.3       TH.data_1.0-10      digest_0.6.25      
 [11] foreach_1.5.0       htmltools_0.4.0     lars_1.2            gdata_2.18.0        fansi_0.4.1        
 [16] magrittr_1.5        checkmate_2.0.0     cluster_2.1.0       mixtools_1.2.0      ROCR_1.0-7         
 [21] modelr_0.1.8        R.utils_2.9.2       sandwich_2.5-1      colorspace_1.4-1    ggrepel_0.8.2      
 [26] rvest_0.3.5         haven_2.3.0         xfun_0.14           crayon_1.3.4        jsonlite_1.6.1     
 [31] survival_3.1-12     zoo_1.8-8           iterators_1.0.12    ape_5.3             glue_1.4.1         
 [36] gtable_0.3.0        kernlab_0.9-29      prabclus_2.3-2      BiocGenerics_0.28.0 DEoptimR_1.0-8     
 [41] scales_1.1.1        mvtnorm_1.1-0       DBI_1.1.0           bibtex_0.4.2.2      Rcpp_1.0.4.6       
 [46] metap_1.3           dtw_1.21-3          plotrix_3.7-8       riverplot_0.6       htmlTable_1.13.3   
 [51] reticulate_1.15     foreign_0.8-71      bit_1.1-15.2        proxy_0.4-24        mclust_5.4.6       
 [56] SDMTools_1.1-221    Formula_1.2-3       tsne_0.1-3          stats4_3.5.3        htmlwidgets_1.5.1  
 [61] httr_1.4.1          FNN_1.1.3           gplots_3.0.3        RColorBrewer_1.1-2  fpc_2.2-5          
 [66] acepack_1.4.1       TFisher_0.2.0       modeltools_0.2-23   ellipsis_0.3.1      ica_1.0-2          
 [71] pkgconfig_2.0.3     R.methodsS3_1.8.0   flexmix_2.3-15      nnet_7.3-14         dbplyr_1.4.3       
 [76] reshape2_1.4.4      tidyselect_1.1.0    rlang_0.4.6         munsell_0.5.0       cellranger_1.1.0   
 [81] tools_3.5.3         cli_2.0.2           generics_0.0.2      broom_0.5.6         ggridges_0.5.2     
 [86] knitr_1.28          bit64_0.9-7         fs_1.4.1            fitdistrplus_1.1-1  robustbase_0.93-6  
 [91] caTools_1.17.1.2    RANN_2.6.1          packrat_0.5.0       pbapply_1.4-2       nlme_3.1-148       
 [96] R.oo_1.23.0         xml2_1.3.2          compiler_3.5.3      rstudioapi_0.11     png_0.1-7          
[101] reprex_0.3.0        stringi_1.4.6       lattice_0.20-41     multtest_2.38.0     vctrs_0.3.0        
[106] mutoss_0.1-12       pillar_1.4.4        lifecycle_0.2.0     lmtest_0.9-37       Rdpack_0.11-1      
[111] irlba_2.3.3         bitops_1.0-6        gbRd_0.4-11         R6_2.4.1            latticeExtra_0.6-28
[116] KernSmooth_2.23-17  gridExtra_2.3       codetools_0.2-16    MASS_7.3-51.6       gtools_3.8.2       
[121] assertthat_0.2.1    withr_2.2.0         mnormt_1.5-7        multcomp_1.4-13     diptest_0.75-7     
[126] parallel_3.5.3      doSNOW_1.0.18       hms_0.5.3           grid_3.5.3          rpart_4.1-15       
[131] class_7.3-17        segmented_1.1-0     Rtsne_0.15          numDeriv_2016.8-1.1 Biobase_2.42.0     
[136] lubridate_1.7.8     base64enc_0.1-3  
```

**Comment by cgao90** [comment]
Hi Sam, 

Thank you for your interest in online liger and reaching out to us regarding those problems.

We've been working on this project and making modifications. We had an update on online liger yesterday and now the plotting functions (including `plotGene` and `plotGeneViolin`) support liger object generated from h5 files. Here is the link to the latest tutorial [()](http://htmlpreview.github.io/?https://github.com/MacoskoLab/liger/blob/online/vignettes/online_iNMF_tutorial.html).

Currently, online liger does not support Seurat-Liger object conversion. For online learning, the liger object needs to be build upon the h5 input from the beginning. 

Could you please try to reinstall the online branch and perform the analysis? 

Best,
Chao

**Comment by samuel-marsh** [comment]
Hi Chao,

Thanks for reply!  So the actual dataset integration using `online_iNMF` seems to work great, both using liger object created from Seurat conversion or just directly reading in the barcodes, features, and matrix files to create liger object.  Only the plotting of the online liger object after that seems to be the issue and again that is totally fixed by just copying the plotting functions from the master branch.  So I'm not sure I understand the requirement for online to be based on H5 input.

Unfortunately for the datasets I'm currently working with I don't have access to the H5 outputs from cell ranger as they are public datasets and the posted files are the features, barcodes, and matrix files (outside of completely reprocessing the files through cell ranger myself). 

I will try and download the latest online branch and run again but I downloaded updates as of last Friday so I believe they are pretty current (minus any changes made this weekend).

Thanks!
Sam

**Comment by jw156605** [comment]
The plotting functions in the online branch previously did not work if you created a liger object from HDF5 files. This is because the functions were implemented to use data stored in memory. We have now fixed this so that you can use the plotting functions with either an object constructed from HDF5 files or an object with all data stored in memory.

**Comment by samuel-marsh** [comment]
Hi Josh,

Thanks for the clarification.  I re-ran analysis yesterday morning with the then current branch and still had the issue.  I saw this morning though that @cgao90 pushed commit to `plotGene` yesterday so I re-ran again this morning with update to newest commits and happy to say that that now plotGene functions properly with in memory object on the online branch.

I'll leave it up to you as to whether you want to close this issue now or leave it open (fine by me to close) but also wanted to let you know as fyi for future changes many of the other plotting functions though appear to still have errors for in memory objects.  `plotGeneViolin()` using branch updated this morning yields the following error:
```
> p2 <- plotGeneViolin(object = object_name, gene = "CX3CR1", return.plots = TRUE)
Error in object@h5file.info[[1]] : subscript out of bounds
```

I would also add that while the PR I pushed for updating the `plotGene` and `plotGenes` functions fine with in memory objects with the online branch it will need updates to function with HDF5-based liger objects too but as I mentioned in PR I think is definitely nice option to have while plotting.

Thanks very much!
Sam
